src/boat_simulator/boat_simulator/nodes/low_level_control/controller.py

```diff
@@ -0,0 +1,260 @@
+            `time_step` (Scalar): Time per iteration given in degrees/second.
+            `kp` (Scalar): Proportional constant when calculating error.
+            `cp` (Scalar): Timeseries of error values computed over time.
+            `max_angle_range` (tuple): Tuple of minimum to maximum control angle range in degrees
+            `rudder_speed` (Scalar): Speed of controller change in degrees per second
+
+        """
+        super().__init__(current_control_ang, time_step, max_angle_range)
+
+        self.current_heading = current_heading
+        self.desired_heading = desired_heading
+        self.kp = kp
+        self.cp = cp
+        self.max_angle_range = max_angle_range
+        self.rudder_speed = rudder_speed
+        self.running_error = 0.0  # keeps track of remaining error between desired and current
+
+    def compute_error(
+        self, desired: Scalar, current: Scalar
+    ) -> Scalar:  # angle passed in as radians
+        """Computes the error between desired and current heading
+        implementation taken from: https://stackoverflow.com/a/2007279
+        Angles are bound with the convention (-180, 180]
+
+        Args:
+            `desired` (Scalar): New desired heading in degrees
+            `current` (Scalar): New current heading in degrees
+
+        Returns:
+            Scalar: The error between the given headings in radians.
+        """
+        current_bound, desired_bound = map(bound_to_180, (current, desired))
+        desired_rad, current_rad = np.deg2rad(desired_bound), np.deg2rad(current_bound)
+
+        error = atan2(sin(desired_rad - current_rad), cos(desired_rad - current_rad))
+
+        return error
+
+    def compute_error_angle(self) -> Scalar:
+        """Computes the feedback angle from the given error (in radians)
+        Calculations done using Raye's implementation
+            https://github.com/UBCSailbot/raye-boat-controller/blob/master/python/tack_controller.py
+
+        Returns:
+            Scalar: Difference between current and target rudder angle in degrees.
+        """
+        error = self.compute_error(self.current_heading, self.desired_heading)
+
+        if abs(error) > (pi / 4):  # limited between (-45, 45)
+            raise ValueError("Heading_error must be between -pi/4 and pi/4")
+
+        rudder_change = (self.kp * error) / (1 + (self.cp * error))
+
+        if abs(rudder_change) > (pi / 2):  # greatest change is 90 degrees
+            raise ValueError("Rudder_change cannot be greater than pi / 2")
+
+        rudder_change_deg = np.rad2deg(rudder_change)
+        self.running_error = rudder_change_deg
+
+        return rudder_change_deg
+
+    def compute_setpoint(self) -> Scalar:
+        """Calculates the rudder setpoint angle. Only called at the very beginning once
+        to determine the setpoint angle.
+
+        Returns:
+            The rudder setpoint angle in degrees
+        """
+        return self.current_control_ang + self.compute_error_angle()
```

Comment:
Apologies but I misinterpreted the paper, but the setpoint the just the value returned by `self.compute_error_angle`. So really, the value returned is $\phi_{\text{setpoint}}$ and not $\Delta\phi$. I realized this after seeing the following in Raye's codebase (and rereading section 5.1 of the paper):

https://github.com/UBCSailbot/raye-boat-controller/blob/3a9a363eb9d769c23a73e9a07666e1cdf28a077a/python/boat_controller_node.py#L103-L105

Thankfully this fix is simple with a few variables and function renaming with some slight changes to documentation. Sorry about this, I am going to update the issue accordingly.

If you plot out the function in desmos as well, and convert to degrees (with $k_p = c_p = 0.3$), you'll see the function saturate at about $\pm 55 \degree$, which makes more sense to support that this is the setpoint rather than the change in the angle.

Reply:
Sounds good I will just take the outpoint of one of my function as the new setpoint! Will take a look at the implementation to confirm.

Reply:
This function would no longer exist for the sail actuator since we pass in a target angle right?